### DISCUSSION TITLE
LDAP (again) or SSO

### DISCUSSION THREAD

**Discussion by keval6b** [dani-garcia/vaultwarden#2396] [Ideas]
Hi all, I'm another advocate for full LDAP support.

As a preface before you immediately dismiss this, I have read all the previous PRs and discussions on the matter. While my understanding is still minimal, I've often found that a new set of eyes can spark a yet unseen path of action. Hopefully my ideas can do so but I understand that there is still a low chance of this.

So from my understanding, the main issue here is that the _Master Password_ is used for encryption and decryption of the vault.  
Therefore, using an LDAP password for vault encryption is impossible since they can change at which point the user is no longer able to decrypt the vault.

---

<details>
<summary>My brain fart process</summary>
There were some solutions posted in the previous discussions, one of which caught my eye:

**Store a generated master password as an ldap attribute on each user**

Now the issue with doing that is that anyone with access to the ldap server, then has the master password of all users. Obviously bad.

So now we can try to encrypt the master password while it's on ldap such that only vaultwarden can decrypt it using a key of some sort.

So then you realise that then anyone with access to vaultwarden has access to the master passwords.

Now we're in a loop.

How to exit the loop?

Impossible

The only person with the knowledge to unlock the vault, even by proxy, must be the user.
</details>

---

So what if we just ask for the old password if we believe that it has changed?

We can figure out if it has changed by storing an old password hash on ldap as an attribute (or within vaultwarden) and then on failure to decrypt, check the old password and re-encrypt the vault with the new password.

I can't see any huge issues with this other than perhaps the old password being multiple generations old (store a list?) or the user having no knowledge of the old password (any ideas?)

**Comment by keval6b** [comment]
Alternately we just enforce that the user's password is always the same as their ldap one.

When they try to login to bitwarden with a working password but we cannot log them in via ldap, we get them to change their bitwarden password such that it does allow login to ldap too?

**Comment by BlackDex** [comment]
In my opinion this will deviate too much from upstream, and it will cause issues in the long run i think.
Also, this is more an SSO kinda solution, which then still is no _full ldap_ of course, and still requires for users to have a separate password to decrypt there vault. But that is just how Bitwarden works, it needs the original password, no hash or whatever which could change if a salt changes even when the password will be the same.

In the end, i think you will only lock-out people instead of providing an easy login.

**Reply by keval6b** [reply]
Would the latter (same password enforcement) not be just a layer on top of the existing implementation?

I don't see much overlap apart from a small amount of logic when logging in and a password reset policy which could be implemented in a generic mannner. I.e. admins could force users to change passwords once every 6 months or something

**Comment by keval6b** [comment]
How does the SSO solution improve this situation. What needs to happen for someone to implement SSO? (I might give it a shot)

**Comment by BlackDex** [comment]
See #1955
There is already an attempt.
